### ISSUE TITLE
LightGBMClassifier failed with ConnectionRefused spark in spark-submit with cluster mode

### ISSUE THREAD

**Issue by ocworld** [Azure/mmlspark#675]
**Describe the bug**
My application based on mmlspark is failed that is launched with spark cluster mode.
Training small dataset is success. However, it is failure to train large dataset.
Spark-submit with client has not any problem. This error, ConnectionRefused, is occurred in cluster mode.

**To Reproduce**
1. An application that is train large dataset with LightGBMClassifier is launched with spark cluster mode

**Expected behavior**
Training is successful.
However. ConnectionRefused Exception is occurred.

**Info (please complete the following information):**
 - MMLSpark Version: v0.17
 - Spark Version: 2.3.0
 - Spark Platform: EMR

** Stacktrace**
```
Please post the stacktrace here if applicable
```
User class threw exception: org.apache.spark.SparkException: Job aborted due to stage failure: Task 0 in stage 4.0 failed 4 times, most recent failure: Lost task 0.3 in stage 4.0 (TID 256, hdp-d08.my.cluster.com, executor 5): java.net.ConnectException: Connection refused (Connection refused)
at java.net.PlainSocketImpl.socketConnect(Native Method)
at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
at java.net.Socket.connect(Socket.java:589)
at java.net.Socket.connect(Socket.java:538)
at java.net.Socket.<init>(Socket.java:434)
at java.net.Socket.<init>(Socket.java:211)
at com.microsoft.ml.spark.TrainUtils$.getNetworkInitNodes(TrainUtils.scala:277)
at com.microsoft.ml.spark.TrainUtils$$anonfun$7.apply(TrainUtils.scala:340)
at com.microsoft.ml.spark.TrainUtils$$anonfun$7.apply(TrainUtils.scala:335)
at com.microsoft.ml.spark.StreamUtilities$.using(StreamUtilities.scala:29)
at com.microsoft.ml.spark.TrainUtils$.trainLightGBM(TrainUtils.scala:334)
at com.microsoft.ml.spark.LightGBMBase$$anonfun$4.apply(LightGBMBase.scala:49)
at com.microsoft.ml.spark.LightGBMBase$$anonfun$4.apply(LightGBMBase.scala:49)
at org.apache.spark.sql.execution.MapPartitionsExec$$anonfun$5.apply(objects.scala:188)
at org.apache.spark.sql.execution.MapPartitionsExec$$anonfun$5.apply(objects.scala:185)
at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$25.apply(RDD.scala:830)
at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$25.apply(RDD.scala:830)
at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
at org.apache.spark.scheduler.Task.run(Task.scala:109)
at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:345)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
at java.lang.Thread.run(Thread.java:745)
Driver stacktrace:
at org.apache.spark.scheduler.DAGScheduler.org$apache$spark$scheduler$DAGScheduler$$failJobAndIndependentStages(DAGScheduler.scala:1599)
at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1587)
at org.apache.spark.scheduler.DAGScheduler$$anonfun$abortStage$1.apply(DAGScheduler.scala:1586)
at scala.collection.mutable.ResizableArray$class.foreach(ResizableArray.scala:59)
at scala.collection.mutable.ArrayBuffer.foreach(ArrayBuffer.scala:48)
at org.apache.spark.scheduler.DAGScheduler.abortStage(DAGScheduler.scala:1586)
at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:831)
at org.apache.spark.scheduler.DAGScheduler$$anonfun$handleTaskSetFailed$1.apply(DAGScheduler.scala:831)
at scala.Option.foreach(Option.scala:257)
at org.apache.spark.scheduler.DAGScheduler.handleTaskSetFailed(DAGScheduler.scala:831)
at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.doOnReceive(DAGScheduler.scala:1820)
at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1769)
at org.apache.spark.scheduler.DAGSchedulerEventProcessLoop.onReceive(DAGScheduler.scala:1758)
at org.apache.spark.util.EventLoop$$anon$1.run(EventLoop.scala:48)
at org.apache.spark.scheduler.DAGScheduler.runJob(DAGScheduler.scala:642)
at org.apache.spark.SparkContext.runJob(SparkContext.scala:2034)
at org.apache.spark.SparkContext.runJob(SparkContext.scala:2131)
at org.apache.spark.rdd.RDD$$anonfun$reduce$1.apply(RDD.scala:1029)
at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:151)
at org.apache.spark.rdd.RDDOperationScope$.withScope(RDDOperationScope.scala:112)
at org.apache.spark.rdd.RDD.withScope(RDD.scala:363)
at org.apache.spark.rdd.RDD.reduce(RDD.scala:1011)
at org.apache.spark.sql.Dataset.reduce(Dataset.scala:1610)
at com.microsoft.ml.spark.LightGBMBase$class.train(LightGBMBase.scala:51)
at com.microsoft.ml.spark.LightGBMClassifier.train(LightGBMClassifier.scala:23)
at com.microsoft.ml.spark.LightGBMClassifier.train(LightGBMClassifier.scala:23)
at org.apache.spark.ml.Predictor.fit(Predictor.scala:118)
at com.my.myapp.infrastructure.MyAppTrainer.train(MyAppTrainer.scala:--)
at com.my.myapp.application.MyAppTrainer.train(MyAppTrainer.scala:--)
at com.my.myapp.myApp$.main(MyAppTrainerApp.scala:--)
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.lang.reflect.Method.invoke(Method.java:498)
at com.my.myapp.util.spark.SparkApp$$anonfun$main$2.apply(SparkApp.scala:--)
at com.my.myapp.util.spark.SparkApp$$anonfun$main$2.apply(SparkApp.scala:--)
at scala.Option.foreach(Option.scala:257)
at com.my.myapp.util.spark.SparkApp$class.main(SparkApp.scala:--)
at com.my.myapp.infrastructure.MyApp$.main(MyApp.scala:--)
at com.my.myapp.MyApp.main(MyApp.scala)
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
at java.lang.reflect.Method.invoke(Method.java:498)
at org.apache.spark.deploy.yarn.ApplicationMaster$$anon$4.run(ApplicationMaster.scala:721)
Caused by: java.net.ConnectException: Connection refused (Connection refused)
at java.net.PlainSocketImpl.socketConnect(Native Method)
at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
at java.net.Socket.connect(Socket.java:589)
at java.net.Socket.connect(Socket.java:538)
at java.net.Socket.<init>(Socket.java:434)
at java.net.Socket.<init>(Socket.java:211)
at com.microsoft.ml.spark.TrainUtils$.getNetworkInitNodes(TrainUtils.scala:277)
at com.microsoft.ml.spark.TrainUtils$$anonfun$7.apply(TrainUtils.scala:340)
at com.microsoft.ml.spark.TrainUtils$$anonfun$7.apply(TrainUtils.scala:335)
at com.microsoft.ml.spark.StreamUtilities$.using(StreamUtilities.scala:29)
at com.microsoft.ml.spark.TrainUtils$.trainLightGBM(TrainUtils.scala:334)
at com.microsoft.ml.spark.LightGBMBase$$anonfun$4.apply(LightGBMBase.scala:49)
at com.microsoft.ml.spark.LightGBMBase$$anonfun$4.apply(LightGBMBase.scala:49)
at org.apache.spark.sql.execution.MapPartitionsExec$$anonfun$5.apply(objects.scala:188)
at org.apache.spark.sql.execution.MapPartitionsExec$$anonfun$5.apply(objects.scala:185)
at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$25.apply(RDD.scala:830)
at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$25.apply(RDD.scala:830)
at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:38)
at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:87)
at org.apache.spark.scheduler.Task.run(Task.scala:109)
at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:345)
at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
at java.lang.Thread.run(Thread.java:745)

If the bug pertains to a specific feature please tag the appropriate [CODEOWNER](https://github.com/Azure/mmlspark/blob/master/CODEOWNERS) for better visibility

**Additional context**
I've tried to increase timeout value. However, in cluster mode, the error is reproduced.

**Comment by imatiach-msft** [comment]
@ocworld the connection refused is usually a red herring and it means that one of the workers failed or succeeded early.  How many rows/cols does your dataset have?  Is it unbalanced data and what is the class distribution?  How many nodes do you have in your cluster and what are the specs?

**Comment by ocworld** [comment]
This connection refused error is occurred because of lack of resources on EMR not succeeded early. The number of features are over 300. I tested LightGBM library on python directly on a local server. The amount of memory is not needed to large as mmlspark. I Have a question about why mmlspark consume large memory than using LightGBM library directly?

I try to do repartition(coalesce) on my private spark cluster, it is not need to large memory as it. It is difficult to understand what operations are need to more memory.

Why more memory is needed to use LightGBM on mmlspark?

**Comment by imatiach-msft** [comment]
@ocworld sorry about the error you are encountering.  Would you be able to share the dataset schema and how many rows the dataset has (I understand it has 300 columns).  Also please share the number of nodes (VMs) and total memory on the cluster, as well as the number of executors.  Just to confirm, you are not using dynamic allocation, is that correct?

**Comment by yinshurman** [comment]
same problem with mmlspark 1.0.0-rc1 and spark 2.4.0-cdh:
`19/10/28 11:31:40 INFO executor.CoarseGrainedExecutorBackend: Got assigned task 4683
19/10/28 11:31:40 INFO executor.Executor: Running task 20.1 in stage 56.0 (TID 4683)
19/10/28 11:31:40 INFO storage.BlockManager: Found block rdd_107_74 remotely
19/10/28 11:31:40 INFO lightgbm.LightGBMClassifier: Successfully bound to port 12424
19/10/28 11:31:40 INFO lightgbm.LightGBMClassifier: LightGBM worker connecting to host: 10.0.12.185 and port: 39301
19/10/28 11:31:40 ERROR executor.Executor: Exception in task 20.1 in stage 56.0 (TID 4683)
java.net.ConnectException: Connection refused (Connection refused)
	at java.net.PlainSocketImpl.socketConnect(Native Method)
	at java.net.AbstractPlainSocketImpl.doConnect(AbstractPlainSocketImpl.java:350)
	at java.net.AbstractPlainSocketImpl.connectToAddress(AbstractPlainSocketImpl.java:206)
	at java.net.AbstractPlainSocketImpl.connect(AbstractPlainSocketImpl.java:188)
	at java.net.SocksSocketImpl.connect(SocksSocketImpl.java:392)
	at java.net.Socket.connect(Socket.java:589)
	at java.net.Socket.connect(Socket.java:538)
	at java.net.Socket.<init>(Socket.java:434)
	at java.net.Socket.<init>(Socket.java:211)
	at com.microsoft.ml.spark.lightgbm.TrainUtils$.getNetworkInitNodes(TrainUtils.scala:299)
	at com.microsoft.ml.spark.lightgbm.TrainUtils$$anonfun$12.apply(TrainUtils.scala:372)
	at com.microsoft.ml.spark.lightgbm.TrainUtils$$anonfun$12.apply(TrainUtils.scala:367)
	at com.microsoft.ml.spark.core.env.StreamUtilities$.using(StreamUtilities.scala:28)
	at com.microsoft.ml.spark.lightgbm.TrainUtils$.trainLightGBM(TrainUtils.scala:366)
	at com.microsoft.ml.spark.lightgbm.LightGBMBase$$anonfun$6.apply(LightGBMBase.scala:145)
	at com.microsoft.ml.spark.lightgbm.LightGBMBase$$anonfun$6.apply(LightGBMBase.scala:145)
	at org.apache.spark.sql.execution.MapPartitionsExec$$anonfun$5.apply(objects.scala:188)
	at org.apache.spark.sql.execution.MapPartitionsExec$$anonfun$5.apply(objects.scala:185)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:836)
	at org.apache.spark.rdd.RDD$$anonfun$mapPartitionsInternal$1$$anonfun$apply$24.apply(RDD.scala:836)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.rdd.MapPartitionsRDD.compute(MapPartitionsRDD.scala:52)
	at org.apache.spark.rdd.RDD.computeOrReadCheckpoint(RDD.scala:324)
	at org.apache.spark.rdd.RDD.iterator(RDD.scala:288)
	at org.apache.spark.scheduler.ResultTask.runTask(ResultTask.scala:90)
	at org.apache.spark.scheduler.Task.run(Task.scala:121)
	at org.apache.spark.executor.Executor$TaskRunner$$anonfun$11.apply(Executor.scala:407)
	at org.apache.spark.util.Utils$.tryWithSafeFinally(Utils.scala:1360)
	at org.apache.spark.executor.Executor$TaskRunner.run(Executor.scala:413)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
19/10/28 11:31:40 INFO executor.CoarseGrainedExecutorBackend: Driver commanded a shutdown
19/10/28 11:31:40 INFO memory.MemoryStore: MemoryStore cleared
19/10/28 11:31:40 INFO storage.BlockManager: BlockManager stopped
19/10/28 11:31:41 INFO util.ShutdownHookManager: Shutdown hook called`

**Comment by ocworld** [comment]
@yinshurman @imatiach-msft I've tried to change default GC to G1GC. After that, less memory is needed to train a model.

A document about AWS EMR is helpful to me.
https://aws.amazon.com/ko/blogs/big-data/best-practices-for-successfully-managing-memory-for-apache-spark-applications-on-amazon-emr/

@imatiach-msft The number of rows and columns are very various. The count of rows is over a hundred million.

**Comment by ocworld** [comment]
I assume that generateSparseDataset needs very huge memory. I've traced where is a memory peak point with additional logs I appended.

Additionally, is there more methods for less required memory?

**Comment by ocworld** [comment]
below code in LightGBMBooster::translate is very expensive.

```scala
val rows = inputRows.toArray
```

As estimating memory by using java.lang.Runtime, more memory is used after the line than before.

**Comment by imatiach-msft** [comment]
@ocworld interesting find, I need to convert the spark dataframe to a lightgbm dataset in memory.  I wonder if there might be a more efficient way to do the conversion?  Also, what version of mmlspark are you running - I had a bugfix where I only selected the relevant columns (label, features) to reduce memory usage, otherwise all columns were being generated in memory that existed on the dataframe, including ones that were not relevant.

**Comment by imatiach-msft** [comment]
"I assume that generateSparseDataset needs very huge memory. I've traced where is a memory peak point with additional logs I appended."
it just generates the native lightgbm in-memory representation for sparse format.  Was the original dataset sparse or dense?  One problem might be that spark can have mixed sparse/dense row vectors, but lightgbm only has a sparse or dense representation, so I take the first row and depending on whether it is sparse or dense convert to the lightgbm representation.  There is definitely a more efficient way to do this, since in the worst case scenario the first row can be sparse and the rest of the rows dense, but there is a tradeoff because then you need to use more compute time to figure out what is the most efficient in-memory representation.

**Comment by ocworld** [comment]
@imatiach-msft Thank you for your response.

Firstly, 1.0.0-rc1 is more efficient as aspect of memory I tested. I used 0.17 and 0.18.1. Now, the package is updated to 1.0.0-rc1. I've tested training on the new version using a dataset that is also tested on old versions. In the same cluster specs and spark parameters as versions raising exception, the new version, 1.0.0-rc1, do not occur error. I don't know exactly different between these versions.

Secondly, I think that a big memory consumer is not generateSparseDataset but inputRows.toArray in LightGBMBooster::translate. When first time, I've checked only committed logs. After that, I've appended more logs for estimating used memory. For calculating it, java.lang.Runtime is imported to use total memory and free memory. In my dataset, after collecting about 6~7GB parquet files on a worker as array, memory usage is increasing over 40GB. This memory usage is already high before generateSparseDataset.

When I tested LightGBM on python, I use a dataset as binary file or svm format. I think, so, it is more efficient than mmlspark as aspect of memory. Is it possible for LightGBM to support stream data or convert rdd to one or more binary format data structure?

**Comment by ocworld** [comment]
In addition, when assign executor memory, it should be considered that is spark.memory.fraction.

**Comment by imatiach-msft** [comment]
@ocworld great questions.  This was the main PR that reduced memory usage which only selected the relevant columns needed:
https://github.com/Azure/mmlspark/pull/586
Version 0.17 did not contain this fix, it was in 0.18.  Without this fix any temporary columns used to create the features column (or other columns not used by lightgbm) were being generated and kept in memory which for some users was causing OOM errors.
LightGBM requires all of the data to be in-memory during training, it does not support a streaming API.  Hence, we need to store the original data in spark and also translate to native dataset representation that LightGBM can train on.  Before translating to native, I do inputRows.toArray, which increases memory usage even more (it seems that inputRows copies the actual spark dataframe).  I expect there to be roughly 3X memory usage of the original data.  However, I think we might be able to release rows after inputRows.toArray as we build up the native dataset.  I don't see how I could build up the native data with inputRows streamed (and not call .toArray) unless I get the length of the inputRows ahead of time and pre-allocate the native arrays, but calling size on the iterator goes through the iterator and I would need to re-iterate over the inputRows again (which I believe I might only be able to do with another mapPartitions call?).

**Comment by KeertiBafna** [comment]
@imatiach-msft I still face this issue in version 0.18.1 . Is there any particular fix or solutionrelated to this ?
